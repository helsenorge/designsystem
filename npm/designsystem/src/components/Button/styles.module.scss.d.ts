export type Styles = {
  button: string;
  button__arrow: string;
  'button__arrow--both-icons': string;
  'button__both-icons': string;
  'button__left-icon': string;
  'button__only-icon': string;
  'button__right-icon': string;
  'button__right-unicode-arrow': string;
  button__text: string;
  'button__text--centered-mobile': string;
  'button__text--ellipsis': string;
  'button--arrow': string;
  'button--arrow--both-icons': string;
  'button--borderless': string;
  'button--both-icons': string;
  'button--destructive': string;
  'button--large': string;
  'button--left-icon': string;
  'button--normal': string;
  'button--on-dark': string;
  'button--only-icon': string;
  'button--outline': string;
  'button--right-icon': string;
  'button-wrapper': string;
  'button-wrapper--fluid': string;
  diagonal: string;
  diagonal__line: string;
  'diagonal--on-dark': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
