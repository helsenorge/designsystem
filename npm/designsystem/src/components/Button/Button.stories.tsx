import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Button from './Button';
import Icon from '../Icons';
import Dog from '../Icons/Dog';
import VerticalDots from '../Icons/VerticalDots';
import Title from '../Title';
import { getColor } from '../../theme/currys/color';
import GridExample from '../GridExample';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    children: {
      control: 'text',
      defaultValue: 'Button',
    },
    ellipsis: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
    fluid: {
      control: 'boolean',
      defaultValue: false,
    },
    htmlMarkup: {
      control: 'select',
      options: ['button', 'a'],
      defaultValue: 'button',
    },
    href: {
      control: 'text',
      defaultValue: 'https://www.helsenorge.no',
    },
  },
} as ComponentMeta<typeof Button>;

const myRef1: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef2: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef3: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef4: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef5: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef6: React.RefObject<HTMLButtonElement> = React.createRef();
const myRef7: React.RefObject<HTMLButtonElement> = React.createRef();

export const Default: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Button {...args} onClick={action('Button clicked')}>
      {args.children}
    </Button>
  </GridExample>
);

export const Concepts: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Concepts'}
    </Title>
    <Button {...args} onClick={action('Button clicked')} concept={'normal'}>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} onClick={action('Button clicked')} concept={'destructive'}>
      {args.children}
    </Button>
  </GridExample>
);

export const Icons: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Icons'}
    </Title>
    <Button {...args} ref={myRef1} onClick={action('Button clicked')}>
      <Icon svgIcon={Dog} />
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} ref={myRef2} onClick={action('Button clicked')}>
      {args.children}
      <Icon svgIcon={Dog} />
    </Button>
    <br />
    <br />
    <Button {...args} ref={myRef3} onClick={action('Button clicked')} arrow>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} ref={myRef4} onClick={action('Button clicked')} arrow>
      <Icon svgIcon={Dog} />
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} ariaLabel={'Verticaldots button'} ref={myRef5} onClick={action('Button clicked')}>
      <Icon svgIcon={VerticalDots} />
    </Button>
  </GridExample>
);

export const Sizes: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Sizes'}
    </Title>
    <Button size={'medium'} {...args} ref={myRef6} onClick={action('Button clicked')}>
      {args.children}
    </Button>
    <br />
    <br />
    <Button size={'large'} {...args} ref={myRef7} onClick={action('Button clicked')}>
      {args.children}
      <Icon svgIcon={Dog} />
    </Button>
  </GridExample>
);

export const Variants: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Variants'}
    </Title>
    <Button {...args} onClick={action('Button clicked')} variant={'fill'}>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} onClick={action('Button clicked')} variant={'outline'}>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} onClick={action('Button clicked')} variant={'borderless'}>
      {args.children}
    </Button>
  </GridExample>
);

export const DarkMode: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'DarkMode'}
    </Title>
    <div style={{ padding: '3rem', backgroundColor: getColor('blueberry', 600) }}>
      <Button {...args} onClick={action('Button clicked')} variant={'fill'} mode={'ondark'}>
        {args.children}
      </Button>
      <br />
      <br />
      <Button {...args} onClick={action('Button clicked')} mode={'ondark'}>
        <Icon svgIcon={Dog} />
        {args.children}
      </Button>
      <br />
      <br />
      <Button {...args} onClick={action('Button clicked')} variant={'outline'} mode={'ondark'}>
        {args.children}
      </Button>
      <br />
      <br />
      <Button {...args} onClick={action('Button clicked')} variant={'borderless'} mode={'ondark'}>
        {args.children}
      </Button>
      <br />
      <br />
      <Button {...args} onClick={action('Button clicked')} variant={'borderless'} mode={'ondark'}>
        <Icon svgIcon={Dog} />
        {args.children}
      </Button>
    </div>
  </GridExample>
);

export const Disabled: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Disabled'}
    </Title>
    <Button {...args} onClick={action('Button clicked')} variant={'fill'} disabled>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} onClick={action('Button clicked')} variant={'outline'} disabled>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} onClick={action('Button clicked')} variant={'borderless'} disabled>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} onClick={action('Button clicked')} variant={'borderless'} disabled>
      <Icon svgIcon={Dog} />
      {'Button long text'}
    </Button>
  </GridExample>
);

export const Fluid: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Fluid'}
    </Title>
    <Button {...args} fluid onClick={action('Button clicked')}>
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} fluid onClick={action('Button clicked')}>
      <Icon svgIcon={Dog} />
      {args.children}
    </Button>
    <br />
    <br />
    <Button {...args} fluid onClick={action('Button clicked')}>
      {args.children}
      <Icon svgIcon={Dog} />
    </Button>
    <br />
    <br />
    <Button {...args} arrow fluid onClick={action('Button clicked')}>
      <Icon svgIcon={Dog} />
      {args.children}
    </Button>
  </GridExample>
);

export const Ellipsis: ComponentStory<typeof Button> = (args: any) => (
  <GridExample>
    <Title margin={2} htmlMarkup={'h3'} appearance={'title3'}>
      {'Ellipsis'}
    </Title>
    <Button {...args} ellipsis onClick={action('Button clicked')}>
      {'Alt for mye tekst her'}
    </Button>
    <br />
    <br />
    <Button {...args} ellipsis onClick={action('Button clicked')}>
      <Icon svgIcon={Dog} />
      {'Alt for mye tekst her'}
    </Button>
  </GridExample>
);

export const AllVariants: ComponentStory<typeof Button> = (args: any) => (
  <>
    <Concepts {...args} />
    <Icons {...args} />
    <Sizes {...args} />
    <Variants {...args} />
    <DarkMode {...args} />
    <Disabled {...args} />
    <Fluid {...args} />
    <Ellipsis {...args} />
  </>
);
