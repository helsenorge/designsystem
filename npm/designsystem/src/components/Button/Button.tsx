import React, { AriaAttributes, useEffect, useRef } from 'react';

import classNames from 'classnames';

import Icon, { IconSize } from './../Icon';
import { HTMLButtonProps, HTMLAnchorProps, AnalyticsId } from '../../constants';
import { useBreakpoint } from '../../hooks/useBreakpoint';
import { useHover } from '../../hooks/useHover';
import { BaseIconElement, useIcons } from '../../hooks/useIcons';
import { getColor } from '../../theme/currys/color';
import { breakpoints } from '../../theme/grid';
import { isTest, isProd } from '../../utils/environment';
import ArrowRight from '../Icons/ArrowRight';

import buttonStyles from './styles.module.scss';

export type ButtonConcept = 'normal' | 'destructive';
export type ButtonVariant = 'fill' | 'outline' | 'borderless';
export type ButtonSize = 'medium' | 'large';
export type ButtonMode = 'onlight' | 'ondark';
export type ButtonTags = 'button' | 'a';
export type ButtonArrows = 'icon' | 'accessibility-character';

export interface ButtonProps extends HTMLButtonProps, HTMLAnchorProps, AriaAttributes {
  /** Sets the aria-label of the button, use when the button only includes an icon */
  ariaLabel?: string;
  /** Gives a unique id to the button */
  id?: string;
  /** Sets the content of the button. */
  children: React.ReactNode;
  /** Adds custom classes to the wrapper element. */
  wrapperClassName?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Enables an arrow icon to the right inside the button (Not available in borderless variant) */
  arrow?: ButtonArrows;
  /** Changes the intent of the button. Mostly changes the color profile. */
  concept?: ButtonConcept;
  /** Disables text wrapping and enables ellipsis. */
  ellipsis?: boolean;
  /** Makes the button scale to full width of its parent element. */
  fluid?: boolean;
  /** Changes the underlying element of the button. */
  htmlMarkup?: ButtonTags;
  /** Changes the button colors for different backgrounds. */
  mode?: ButtonMode;
  /** Function that is called when the Button loses focus */
  onBlur?: () => void;
  /** Function that is called when clicked */
  onClick?: (e?: React.MouseEvent<HTMLElement, MouseEvent> | React.FormEvent<{}> | React.KeyboardEvent<HTMLUListElement> | null) => void;
  /** Changes the button colors for different backgrounds. (Large not available in borderless variant) */
  size?: ButtonSize;
  /** Changes the visual representation of the button. */
  variant?: ButtonVariant;
  /** Specifies the focus order relative to the other buttons or controls on the page  */
  tabIndex?: number;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Button type. Default: button */
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

const getIconColor = (
  fill: boolean,
  borderless: boolean,
  disabled: boolean,
  concept: ButtonConcept,
  ondark: boolean,
  mobile: boolean
): string => {
  if (mobile && disabled) {
    return !ondark || fill ? getColor('neutral', !borderless ? 700 : 500) : getColor('white');
  }
  if (disabled) {
    return !ondark || fill ? getColor('neutral', 500) : `${getColor('white')}b3`;
  }
  if ((fill && !ondark) || (!fill && ondark)) {
    return 'white';
  }

  return concept === 'normal' ? getColor('blueberry', 600) : getColor('cherry', 500);
};
const getLargeIconSize = (large: boolean, mobile: boolean): IconSize => {
  if (mobile && large) return IconSize.Small;
  if (large) return IconSize.Medium;
  return IconSize.XSmall;
};

const checkOnlyIconAria = (onlyIcon: boolean, ariaLabel: string | undefined, devEnv: boolean): void => {
  if (devEnv && onlyIcon && (ariaLabel === undefined || ariaLabel === '')) {
    throw new Error('Fyll inn ariaLabel prop på Button uten tekst for å opprettholde UU krav');
  }
};

const Button = React.forwardRef(function ButtonForwardedRef(
  props: ButtonProps,
  ref: React.ForwardedRef<HTMLButtonElement | HTMLAnchorElement>
) {
  const {
    ariaLabel,
    id,
    children,
    wrapperClassName,
    className,
    arrow,
    concept = 'normal',
    disabled = false,
    ellipsis = false,
    fluid = false,
    htmlMarkup = 'button',
    mode = 'onlight',
    onBlur,
    onClick,
    size = 'medium',
    variant = 'fill',
    href,
    tabIndex,
    testId,
    target,
    type = 'button',
    ...restProps
  } = props;

  const [leftIcon, rightIcon, restChildren] = useIcons(React.Children.toArray(children));
  const { hoverRef, isHovered } =
    htmlMarkup === 'button'
      ? useHover<HTMLButtonElement>(ref as React.RefObject<HTMLButtonElement>)
      : useHover<HTMLAnchorElement>(ref as React.RefObject<HTMLAnchorElement>);
  const buttonContentRef = useRef<HTMLDivElement>(null);
  const onlyIcon = !!(leftIcon || rightIcon) && !restChildren;
  const bothIcons = leftIcon && (rightIcon || arrow) && !onlyIcon;
  const onDark = mode === 'ondark';
  const breakpoint = useBreakpoint();
  const mobile = breakpoint < breakpoints.md;
  const destructive = concept === 'destructive' && !disabled;
  const outlineVariant = variant === 'outline';
  const borderlessVariant = variant === 'borderless';
  const iconColor = getIconColor(variant === 'fill', borderlessVariant, disabled, concept, onDark, mobile);
  const hasArrow = arrow === 'icon' && !borderlessVariant;
  const large = size === 'large' && !destructive && !borderlessVariant;
  const hasUURightArrow = arrow === 'accessibility-character' && !fluid && !leftIcon && !rightIcon && !hasArrow && borderlessVariant;
  const rest = { ...restProps };

  const buttonWrapperClasses = classNames(
    buttonStyles['button-wrapper'],
    { [buttonStyles['button-wrapper--fluid']]: fluid || ellipsis },
    wrapperClassName
  );
  const buttonClasses = classNames(
    buttonStyles.button,
    {
      [buttonStyles['button--destructive']]: destructive,
      [buttonStyles['button--normal']]: !large,
      [buttonStyles['button--large']]: large,
      [buttonStyles['button--outline']]: outlineVariant,
      [buttonStyles['button--borderless']]: borderlessVariant,
      [buttonStyles['button--left-icon']]: leftIcon && !onlyIcon,
      [buttonStyles['button--right-icon']]: rightIcon && !onlyIcon,
      [buttonStyles['button--both-icons']]: bothIcons,
      [buttonStyles['button--only-icon']]: onlyIcon,
      [buttonStyles['button--arrow']]: hasArrow,
      [buttonStyles['button--on-dark']]: onDark,
    },
    className
  );
  const buttonTextClasses = classNames(buttonStyles['button__text'], {
    [buttonStyles['button__text--ellipsis']]: ellipsis,
  });
  const diagonalClasses = classNames(buttonStyles['diagonal'], {
    [buttonStyles['diagonal--on-dark']]: onDark,
  });

  useEffect(() => {
    checkOnlyIconAria(onlyIcon, ariaLabel, !isTest() && !isProd());
  }, []);

  const renderIcon = (iconElement: BaseIconElement | null, iconSize: number, iconClassName?: string): BaseIconElement | null => {
    return iconElement
      ? React.cloneElement(iconElement, {
          size: iconSize,
          color: iconElement?.props.color ? iconElement.props.color : iconColor,
          isHovered,
          className: iconClassName,
        })
      : null;
  };

  const renderButtonContent = (): JSX.Element => {
    return (
      <span className={buttonTextClasses} ref={buttonContentRef}>
        {disabled && borderlessVariant && (
          <span className={diagonalClasses}>
            <span className={buttonStyles['diagonal__line']} />
          </span>
        )}
        <span>{onlyIcon ? ariaLabel : restChildren}</span>
      </span>
    );
  };

  const renderbuttonContentWrapper = (): JSX.Element => (
    <span className={buttonClasses}>
      {renderIcon(leftIcon, getLargeIconSize(large, mobile), !onlyIcon ? buttonStyles['button__left-icon'] : undefined)}
      {renderButtonContent()}
      {hasArrow
        ? renderIcon(
            <Icon svgIcon={ArrowRight} />,
            getLargeIconSize(large, mobile),
            classNames(buttonStyles['button__arrow'], { [buttonStyles['button__arrow--both-icons']]: bothIcons })
          )
        : renderIcon(rightIcon, getLargeIconSize(large, mobile), buttonStyles['button__right-icon'])}
      {hasUURightArrow && (
        <span style={{ color: iconColor }} className={buttonStyles['button__right-unicode-arrow']} aria-hidden>
          {'  →'}
        </span>
      )}
    </span>
  );

  return (
    <>
      {htmlMarkup === 'button' && (
        <button
          id={id}
          onBlur={onBlur}
          onClick={onClick}
          disabled={disabled}
          data-testid={testId}
          data-analyticsid={AnalyticsId.Button}
          className={buttonWrapperClasses}
          ref={hoverRef as React.ForwardedRef<HTMLButtonElement>}
          tabIndex={tabIndex}
          type={type}
          {...rest}
        >
          {renderbuttonContentWrapper()}
        </button>
      )}
      {htmlMarkup === 'a' && (
        <a
          id={id}
          onBlur={onBlur}
          onClick={onClick}
          data-testid={testId}
          data-analyticsid={AnalyticsId.Button}
          className={buttonWrapperClasses}
          href={href}
          target={target}
          rel={target === '_blank' ? 'noopener noreferrer' : props.rel}
          ref={hoverRef as React.ForwardedRef<HTMLAnchorElement>}
          tabIndex={tabIndex}
          {...restProps}
        >
          {renderbuttonContentWrapper()}
        </a>
      )}
    </>
  );
});

export default Button;
