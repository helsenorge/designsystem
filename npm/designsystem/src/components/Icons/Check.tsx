import React from 'react';
import { getIcon, SvgPathProps } from './Icon';

const Check: React.FC<SvgPathProps> = ({ size, isHovered }: SvgPathProps): JSX.Element => {
  const normal = <path d="M22.504 31.198l-9.59-9.966 1.441-1.387 8.149 8.468 14.455-15.016 1.441 1.386z" />;
  const normalHover = <path d="M22.504 31.198l-7.816-8.121 1.441-1.387 6.375 6.623 15.859-16.474 1.44 1.386z" />;
  const small = <path d="M22.504 31.578l-9.781-10.162 1.822-1.752 7.959 8.269 14.265-14.819 1.821 1.752z" />;
  const smallHover = <path d="M22.504 31.578l-7.816-8.124 1.82-1.75 5.996 6.229 16.321-16.956 1.821 1.752z" />;

  return getIcon(size, isHovered, normal, normalHover, small, smallHover);
};

export default Check;
