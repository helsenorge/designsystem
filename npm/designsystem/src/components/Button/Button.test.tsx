import React from 'react';

import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi as jest } from 'vitest';

import Button from './Button';
import Icon from '../Icon';
import Check from '../Icons/Check';
import Dog from '../Icons/Dog';

describe('Gitt at button skal vises', (): void => {
  describe('Når button rendres', (): void => {
    test('Så vises button', (): void => {
      const { container } = render(<Button>Button</Button>);
      expect(container).toMatchSnapshot();
    });

    test('Så rendres button uten aria props', (): void => {
      render(
        <div>
          <Button testId={'test01'}>Button</Button>
        </div>
      );

      const testFill = screen.getByTestId('test01');

      expect(testFill).not.toHaveAttribute('aria-label');
      expect(testFill).not.toHaveAttribute('aria-expanded');
    });

    test('Så er text klasse på', (): void => {
      render(<Button>Button text</Button>);

      const testButtonText = screen.getByText('Button text').parentElement;
      expect(testButtonText?.className).toBe('button__text');
    });
  });

  describe('Når button rendres med id', (): void => {
    test('Så blir id satt', (): void => {
      render(
        <Button id="id123" testId={'test01'}>
          Button
        </Button>
      );
      const testFill = screen.getByTestId('test01');

      expect(testFill.id).toBe('id123');
    });
  });

  describe('Når button rendres med ikoner', (): void => {
    test('Så vises ikonene', (): void => {
      const { container } = render(
        <Button>
          <Icon svgIcon={Check} />
          Button
          <Icon svgIcon={Check} />
        </Button>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når button rendres kun med ikon', (): void => {
    test('Så inneholder knappen bare ikon, og ariaLabel kan leses', (): void => {
      const { container } = render(
        <Button ariaLabel="Check me">
          <Icon svgIcon={Check} />
        </Button>
      );

      const button = screen.getByText('Check me');

      expect(button).toBeVisible();
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når button rendres med to ikoner', (): void => {
    test('Så skal button ha to ikoner ', (): void => {
      render(
        <Button arrow="icon" testId="test">
          <Icon svgIcon={Dog} />
          Button
        </Button>
      );

      const text = screen.getByText('Button');
      const button = screen.getByTestId('test');

      expect(text).toBeVisible();
      expect(button.firstElementChild?.className).toBe('button button--normal button--left-icon button--both-icons button--arrow');
    });
  });

  describe('Når button rendres med ellipsis på', (): void => {
    test('Så brukes ellipsis på overflødig tekst', (): void => {
      render(
        <Button ellipsis={true} testId={'test01'}>
          Button text
        </Button>
      );

      const button = screen.getByTestId('test01');
      const testButtonText = screen.getByText('Button text').parentElement;
      expect(button.className).toBe('button-wrapper button-wrapper--fluid');
      expect(testButtonText?.className).toBe('button__text button__text--ellipsis');
    });
  });

  describe('Når button rendres med fluid', (): void => {
    test('Så settes riktige klasse på wrapper', (): void => {
      render(
        <Button testId={'test01'} fluid={true}>
          Button text
          <Icon svgIcon={Check} />
        </Button>
      );

      const buttonWrapper = screen.getByTestId('test01');
      expect(buttonWrapper.className).toBe('button-wrapper button-wrapper--fluid');
    });
  });

  describe('Når button rendres med fluid og textPosition som centered-mobile', (): void => {
    test('Så settes riktige klasse på text', (): void => {
      render(
        <Button testId={'test01'} fluid={true} textPosition={'centered-mobile'}>
          Button text
          <Icon svgIcon={Check} />
        </Button>
      );

      const buttonText = screen.getByText('Button text').parentElement;
      expect(buttonText?.className).toBe('button__text button__text--centered-mobile');
    });
  });

  describe('Når button rendres uten fluid og textPosition som centered-mobile', (): void => {
    test('Så settes riktige klasse på text', (): void => {
      render(
        <Button testId={'test01'} textPosition={'centered-mobile'}>
          Button text
          <Icon svgIcon={Check} />
        </Button>
      );

      const buttonText = screen.getByText('Button text').parentElement;
      expect(buttonText?.className).toBe('button__text');
    });
  });

  describe('Når button rendres som button med onBlur-handler', () => {
    test('Så kalles onBlur-handleren når man klikker på knappen', async () => {
      const handleBlur = jest.fn();

      render(
        <Button onBlur={handleBlur} htmlMarkup="button">
          Lenketekst
        </Button>
      );

      const link = screen.getByRole('button', { name: 'Lenketekst' });

      await userEvent.click(link);
      await userEvent.tab();

      expect(handleBlur).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når button rendres som button med onClick-handler', () => {
    test('Så kalles onClick-handleren når man klikker på knappen', async () => {
      const handleClick = jest.fn();

      render(
        <Button onClick={handleClick} htmlMarkup="button">
          Lenketekst
        </Button>
      );

      const link = screen.getByRole('button', { name: 'Lenketekst' });

      await userEvent.click(link);

      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når button rendres som lenke med onClick-handler', () => {
    test('Så kalles onClick-handleren når man klikker på knappen', async () => {
      const handleClick = jest.fn();

      render(
        <Button onClick={handleClick} htmlMarkup="a" href="#">
          Lenketekst
        </Button>
      );

      const link = screen.getByRole('link', { name: 'Lenketekst' });

      await userEvent.click(link);

      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når button rendres med disabled', (): void => {
    test('Så kan ikke knappen trykkes på', (): void => {
      const handleClick = jest.fn();

      render(
        <Button testId={'test01'} onClick={handleClick} disabled={true}>
          Button
        </Button>
      );

      const testButton = screen.getByTestId('test01');

      fireEvent.click(testButton);

      expect(handleClick).toHaveBeenCalledTimes(0);
      expect(testButton).toBeDisabled();
    });
  });

  describe('Når button rendres med concept', (): void => {
    test('Så er background-color riktig', (): void => {
      render(
        <div>
          <Button testId={'test01'} concept={'normal'}>
            Button
          </Button>
          <Button testId={'test02'} concept={'destructive'}>
            Button
          </Button>
        </div>
      );

      const testButton1 = screen.getByTestId('test01').children[0];
      const testButton2 = screen.getByTestId('test02').children[0];

      expect(testButton1.className).toBe('button button--normal');
      expect(testButton2.className).toBe('button button--destructive button--normal');
    });
  });

  describe('Når button rendres med mode ondark', (): void => {
    test('Så er farger riktig', (): void => {
      render(
        <Button testId={'test01'} mode={'ondark'}>
          Button
        </Button>
      );

      const testButton = screen.getByTestId('test01').children[0];

      expect(testButton.className).toBe('button button--normal button--on-dark');
    });
  });

  describe('Når button rendres med size large', (): void => {
    test('Så er ikon riktig størrelse', (): void => {
      render(
        <Button size={'large'}>
          <Icon testId={'test01'} svgIcon={Check} />
          Button
        </Button>
      );

      const testButtonIcon = screen.getByTestId('test01');
      const height = getComputedStyle(testButtonIcon).getPropertyValue('min-height');
      const width = getComputedStyle(testButtonIcon).getPropertyValue('min-width');

      expect(height).toBe('64px');
      expect(width).toBe('64px');
    });
  });

  describe('Når button rendres med varianter', (): void => {
    test('Så rendres innholdet riktig', (): void => {
      render(
        <div>
          <Button testId={'test01'} variant={'fill'}>
            Button
          </Button>
          <Button testId={'test02'} variant={'outline'}>
            Button
          </Button>
          <Button testId={'test03'} variant={'borderless'}>
            Button
          </Button>
        </div>
      );

      const testFill = screen.getByTestId('test01').children[0];

      expect(testFill.className).toBe('button button--normal');

      const testOutline = screen.getByTestId('test02').children[0];

      expect(testOutline.className).toBe('button button--normal button--outline');

      const testBorderless = screen.getByTestId('test03').children[0];

      expect(testBorderless.className).toBe('button button--normal button--borderless');
    });
  });
  describe('Når button rendres med borderless variant', (): void => {
    test('Så er ikke large size tilgjengelig', (): void => {
      render(
        <div>
          <Button testId={'test01'} variant={'borderless'} size={'large'}>
            Button
          </Button>
        </div>
      );

      const testFill = screen.getByTestId('test01').children[0];

      expect(testFill.className).toBe('button button--normal button--borderless');
    });
    test('Så er ikke arrow som icon tilgjengelig', (): void => {
      render(
        <div>
          <Button testId={'test01'} variant={'borderless'} arrow="icon">
            Button
          </Button>
        </div>
      );

      const testFill = screen.getByTestId('test01').children[0];

      expect(testFill.className).toBe('button button--normal button--borderless');
    });
  });
  describe('Når button rendres med aria props', (): void => {
    test('Så rendres innholdet riktig', (): void => {
      render(
        <div>
          <Button testId={'test01'} aria-label={'aria-label-test'} aria-expanded={true}>
            Button
          </Button>
        </div>
      );

      const testFill = screen.getByTestId('test01');

      expect(testFill).toHaveAttribute('aria-label', 'aria-label-test');
      expect(testFill).toHaveAttribute('aria-expanded', 'true');
    });
  });
  describe('Når button rendres som anchor med target=_blank', (): void => {
    test('Så skal den ha riktig rel-attributt', (): void => {
      render(
        <div>
          <Button testId={'tester-rel'} htmlMarkup="a" target="_blank">
            Button
          </Button>
        </div>
      );

      const testFill = screen.getByTestId('tester-rel');

      expect(testFill).toHaveAttribute('rel', 'noopener noreferrer');
      expect(testFill).toHaveAttribute('target', '_blank');
    });
  });
  describe('Gitt at Button vises', (): void => {
    test('Så har den riktig analyticsid', (): void => {
      render(<Button testId={'knapp'}>Tekst</Button>);

      const button = screen.getByTestId('knapp');

      expect(button).toHaveAttribute('data-analyticsid', 'button');
    });
  });
  describe('Når Button ikke har noen type', (): void => {
    test('Så er den av typen button', (): void => {
      render(<Button testId={'knapp'}>Tekst</Button>);

      const button = screen.getByTestId('knapp');

      expect(button).toHaveAttribute('type', 'button');
    });
  });
  describe('Når Button har type submit', (): void => {
    test('Så er den av typen submit', (): void => {
      render(
        <Button testId={'knapp'} type="submit">
          Tekst
        </Button>
      );

      const button = screen.getByTestId('knapp');

      expect(button).toHaveAttribute('type', 'submit');
    });
  });
  describe('Når Button har type arrow er satt til accessibility-character', (): void => {
    test('Så får den → character lagt til', (): void => {
      render(
        <Button testId={'knapp'} arrow="accessibility-character" variant="borderless">
          {'Tekst'}
        </Button>
      );

      const buttonArrow = screen.getByText('→');

      expect(buttonArrow).toBeInTheDocument();
    });
    test('Så er → skjult for skjermlesere', (): void => {
      render(
        <Button testId={'knapp'} arrow="accessibility-character" variant="borderless">
          {'Tekst'}
        </Button>
      );

      const buttonArrow = screen.getByRole('button', { name: 'Tekst' });

      expect(buttonArrow).toBeVisible();
    });
  });
  describe('Når Button har type arrow er satt til accessibility-character og button props som ikke støtter det', (): void => {
    test('Så får den ikke → character lagt til', (): void => {
      render(
        <>
          <Button testId={'knapp'} arrow="accessibility-character" variant="outline">
            {'Tekst'}
          </Button>
          <Button testId={'knapp'} arrow="accessibility-character">
            {'Tekst'}
          </Button>
          <Button testId={'knapp'} arrow="accessibility-character" fluid>
            {'Tekst'}
          </Button>
          <Button testId={'knapp'} arrow="accessibility-character">
            <Icon svgIcon={Check} />
            {'Tekst'}
          </Button>
        </>
      );

      const buttonArrow = screen.queryByText('→');

      expect(buttonArrow).not.toBeInTheDocument();
    });
  });
});
