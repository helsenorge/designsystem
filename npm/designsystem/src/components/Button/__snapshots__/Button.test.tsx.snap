// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at button skal vises > Når button rendres > Så vises button 1`] = `
<div>
  <button
    class="button-wrapper"
    data-analyticsid="button"
    type="button"
  >
    <span
      class="button button--normal"
    >
      <span
        class="button__text"
      >
        <span>
          Button
        </span>
      </span>
    </span>
  </button>
</div>
`;

exports[`Gitt at button skal vises > Når button rendres kun med ikon > Så inneholder knappen bare ikon, og ariaLabel kan leses 1`] = `
<div>
  <button
    class="button-wrapper"
    data-analyticsid="button"
    type="button"
  >
    <span
      class="button button--normal button--only-icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        color="white"
        data-analyticsid="icon"
        fill="white"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M22.504 31.578l-9.781-10.162 1.822-1.752 7.959 8.269 14.265-14.819 1.821 1.752z"
        />
      </svg>
      <span
        class="button__text"
      >
        <span>
          Check me
        </span>
      </span>
    </span>
  </button>
</div>
`;

exports[`Gitt at button skal vises > Når button rendres med ikoner > Så vises ikonene 1`] = `
<div>
  <button
    class="button-wrapper"
    data-analyticsid="button"
    type="button"
  >
    <span
      class="button button--normal button--left-icon button--right-icon button--both-icons"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon button__left-icon"
        color="white"
        data-analyticsid="icon"
        fill="white"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M22.504 31.578l-9.781-10.162 1.822-1.752 7.959 8.269 14.265-14.819 1.821 1.752z"
        />
      </svg>
      <span
        class="button__text"
      >
        <span>
          Button
        </span>
      </span>
      <svg
        aria-hidden="true"
        class="hnds-style-icon button__right-icon"
        color="white"
        data-analyticsid="icon"
        fill="white"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M22.504 31.578l-9.781-10.162 1.822-1.752 7.959 8.269 14.265-14.819 1.821 1.752z"
        />
      </svg>
    </span>
  </button>
</div>
`;
